File: commandline/internal/handler/issues.go
package handler

import (
	"encoding/json"
	"net/http"

	"github.com/briantol/folio/internal/model"
	"github.com/briantol/folio/internal/store"
	"github.com/go-chi/chi/v5"
)

// IssuesHandler handles /api/issues routes.
type IssuesHandler struct {
	issues *store.IssueStore
}

// NewIssuesHandler creates a new IssuesHandler.
func NewIssuesHandler(i *store.IssueStore) *IssuesHandler {
	return &IssuesHandler{issues: i}
}

// Routes returns a chi router for /api/issues.
func (h *IssuesHandler) Routes() chi.Router {
	r := chi.NewRouter()

	r.Get("/", h.list)
	r.Post("/", h.create)
	r.Patch("/reorder", h.reorder)

	r.Route("/{slug}", func(r chi.Router) {
		r.Get("/", h.get)
		r.Put("/", h.update)
		r.Delete("/", h.delete)

		r.Get("/artifacts", h.listArtifacts)
		r.Post("/artifacts/upload", h.uploadArtifact)
		r.Post("/artifacts/create", h.createArtifact)

		r.Route("/artifacts/{filename}", func(r chi.Router) {
			r.Get("/", h.getArtifact)
			r.Put("/", h.saveArtifact)
			r.Delete("/", h.deleteArtifact)
		})
	})

	return r
}

func (h *IssuesHandler) list(w http.ResponseWriter, r *http.Request) {
	page := QueryInt(r, "page", 1, 1, 999999)
	limit := QueryInt(r, "limit", 25, 1, 100)

	var assignee *string
	if v := r.URL.Query().Get("assignee"); v != "" {
		if v == "__unassigned__" {
			empty := ""
			assignee = &empty
		} else {
			assignee = &v
		}
	}

	var feature *string
	if v := r.URL.Query().Get("feature"); v != "" {
		if v == "__unlinked__" {
			empty := ""
			feature = &empty
		} else {
			feature = &v
		}
	}

	params := model.ListIssuesParams{
		Page:      page,
		Limit:     limit,
		Status:    QueryCSV(r, "status"),
		Type:      QueryCSV(r, "type"),
		Priority:  QueryCSV(r, "priority"),
		Assignee:  assignee,
		Feature:   feature,
		PointsMin: QueryFloat(r, "points_min"),
		PointsMax: QueryFloat(r, "points_max"),
		Labels:    QueryCSV(r, "labels"),
		Sort:      r.URL.Query().Get("sort"),
		Dir:       r.URL.Query().Get("dir"),
	}
	if params.Sort == "" {
		params.Sort = "order"
	}
	if params.Dir == "" {
		params.Dir = "asc"
	}

	result := h.issues.List(params)
	JSON(w, 200, result)
}

func (h *IssuesHandler) get(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	issue := h.issues.Get(slug)
	if issue == nil {
		JSON(w, 404, map[string]interface{}{"error": "Issue not found", "slug": slug})
		return
	}
	JSON(w, 200, issue)
}

func (h *IssuesHandler) create(w http.ResponseWriter, r *http.Request) {
	var input model.CreateIssueInput
	if err := ReadJSON(r, &input); err != nil {
		ErrorJSON(w, 400, "Invalid JSON body")
		return
	}

	if input.Title == "" {
		ErrorJSON(w, 422, "title: Title is required")
		return
	}

	issue, err := h.issues.Create(input)
	if err != nil {
		msg := err.Error()
		if containsString(msg, "already exists") || containsString(msg, "file exists") {
			ErrorJSON(w, 409, msg)
			return
		}
		ErrorJSON(w, 500, msg)
		return
	}

	JSON(w, 201, issue)
}

func (h *IssuesHandler) update(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")

	var rawMap map[string]json.RawMessage
	if err := ReadJSON(r, &rawMap); err != nil {
		ErrorJSON(w, 400, "Invalid JSON body")
		return
	}

	var input model.UpdateIssueInput

	if v, ok := rawMap["title"]; ok {
		var s string
		if json.Unmarshal(v, &s) == nil {
			if s == "" {
				ErrorJSON(w, 422, "title: String must contain at least 1 character(s)")
				return
			}
			input.Title = &s
		}
	}
	if v, ok := rawMap["status"]; ok {
		var s string
		if json.Unmarshal(v, &s) == nil {
			input.Status = &s
		}
	}
	if v, ok := rawMap["type"]; ok {
		var s string
		if json.Unmarshal(v, &s) == nil {
			input.Type = &s
		}
	}
	if v, ok := rawMap["priority"]; ok {
		var s string
		if json.Unmarshal(v, &s) == nil {
			input.Priority = &s
		}
	}
	if v, ok := rawMap["assignees"]; ok {
		input.AssigneesSet = true
		json.Unmarshal(v, &input.Assignees)
	}
	if v, ok := rawMap["points"]; ok {
		input.PointsSet = true
		json.Unmarshal(v, &input.Points)
	}
	if v, ok := rawMap["sprint"]; ok {
		input.SprintSet = true
		json.Unmarshal(v, &input.Sprint)
	}
	if v, ok := rawMap["feature"]; ok {
		input.FeatureSet = true
		json.Unmarshal(v, &input.Feature)
	}
	if v, ok := rawMap["labels"]; ok {
		input.LabelsSet = true
		json.Unmarshal(v, &input.Labels)
	}
	if v, ok := rawMap["body"]; ok {
		var s string
		if json.Unmarshal(v, &s) == nil {
			input.Body = &s
		}
	}

	issue, err := h.issues.Update(slug, input)
	if err != nil {
		ErrorJSON(w, 500, err.Error())
		return
	}
	if issue == nil {
		JSON(w, 404, map[string]interface{}{"error": "Issue not found", "slug": slug})
		return
	}

	JSON(w, 200, issue)
}

func (h *IssuesHandler) delete(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	if !h.issues.Delete(slug) {
		JSON(w, 404, map[string]interface{}{"error": "Issue not found", "slug": slug})
		return
	}
	JSON(w, 200, map[string]interface{}{"ok": true, "slug": slug})
}

func (h *IssuesHandler) reorder(w http.ResponseWriter, r *http.Request) {
	var input struct {
		Slugs  []string `json:"slugs"`
		Offset int      `json:"offset"`
	}
	if err := ReadJSON(r, &input); err != nil {
		ErrorJSON(w, 400, "Invalid JSON body")
		return
	}
	if input.Slugs == nil {
		ErrorJSON(w, 422, "slugs: Required")
		return
	}
	h.issues.Reorder(input.Slugs, input.Offset)
	JSON(w, 200, map[string]interface{}{"ok": true})
}

func (h *IssuesHandler) listArtifacts(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	artifacts, ok := h.issues.ListArtifacts(slug)
	if !ok {
		JSON(w, 404, map[string]interface{}{"error": "Issue not found", "slug": slug})
		return
	}
	JSON(w, 200, artifacts)
}

func (h *IssuesHandler) uploadArtifact(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")

	if err := r.ParseMultipartForm(32 << 20); err != nil {
		ErrorJSON(w, 400, "Invalid multipart form")
		return
	}

	file, header, err := r.FormFile("file")
	if err != nil {
		ErrorJSON(w, 400, "file is required")
		return
	}
	defer file.Close()

	filename := header.Filename
	if filename == "" || filename == "ISSUE.md" {
		ErrorJSON(w, 400, "Invalid filename")
		return
	}

	data := make([]byte, header.Size)
	if _, err := file.Read(data); err != nil {
		ErrorJSON(w, 500, "Failed to read file")
		return
	}

	artifact := h.issues.SaveArtifactBuffer(slug, filename, data)
	if artifact == nil {
		JSON(w, 404, map[string]interface{}{"error": "Issue not found", "slug": slug})
		return
	}
	JSON(w, 201, artifact)
}

func (h *IssuesHandler) createArtifact(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	var input struct {
		Filename string `json:"filename"`
	}
	if err := ReadJSON(r, &input); err != nil {
		ErrorJSON(w, 400, "Invalid JSON body")
		return
	}
	if input.Filename == "" {
		ErrorJSON(w, 422, "filename: Filename is required")
		return
	}
	if input.Filename == "ISSUE.md" {
		ErrorJSON(w, 400, "Invalid filename")
		return
	}

	if _, _, ok := h.issues.GetArtifactFilePath(slug, input.Filename); ok {
		JSON(w, 409, map[string]interface{}{"error": "File already exists", "filename": input.Filename})
		return
	}

	artifact := h.issues.SaveArtifactContent(slug, input.Filename, "")
	if artifact == nil {
		JSON(w, 404, map[string]interface{}{"error": "Issue not found", "slug": slug})
		return
	}
	JSON(w, 201, artifact)
}

func (h *IssuesHandler) getArtifact(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	filename := chi.URLParam(r, "filename")
	_, raw := r.URL.Query()["raw"]

	if raw || !store.IsTextArtifact(filename) {
		path, mime, ok := h.issues.GetArtifactFilePath(slug, filename)
		if !ok {
			JSON(w, 404, map[string]interface{}{"error": "Artifact not found", "slug": slug, "filename": filename})
			return
		}
		w.Header().Set("Content-Type", mime)
		w.Header().Set("Cache-Control", "no-cache")
		http.ServeFile(w, r, path)
		return
	}

	content := h.issues.GetArtifactContent(slug, filename)
	if content == nil {
		JSON(w, 404, map[string]interface{}{"error": "Artifact not found", "slug": slug, "filename": filename})
		return
	}
	JSON(w, 200, content)
}

func (h *IssuesHandler) saveArtifact(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	filename := chi.URLParam(r, "filename")

	var input struct {
		Content string `json:"content"`
	}
	if err := ReadJSON(r, &input); err != nil {
		ErrorJSON(w, 400, "Invalid JSON body")
		return
	}

	artifact := h.issues.SaveArtifactContent(slug, filename, input.Content)
	if artifact == nil {
		JSON(w, 404, map[string]interface{}{"error": "Issue not found or invalid filename", "slug": slug, "filename": filename})
		return
	}
	JSON(w, 200, artifact)
}

func (h *IssuesHandler) deleteArtifact(w http.ResponseWriter, r *http.Request) {
	slug := chi.URLParam(r, "slug")
	filename := chi.URLParam(r, "filename")
	if !h.issues.DeleteArtifact(slug, filename) {
		JSON(w, 404, map[string]interface{}{"error": "Artifact not found", "slug": slug, "filename": filename})
		return
	}
	JSON(w, 200, map[string]interface{}{"ok": true, "slug": slug, "filename": filename})
}

File: api/src/routes/issues.ts
import { readFileSync } from "node:fs"
import { Hono } from "hono"
import {
  listIssues,
  getIssue,
  createIssue,
  updateIssue,
  deleteIssue,
  listIssueArtifacts,
  reorderIssues,
  getArtifactContent,
  getArtifactFilePath,
  saveArtifactContent,
  saveArtifactBuffer,
  deleteArtifact,
  isTextArtifact,
} from "../store/issues.js"
import type { IssueStatus, IssueType, Priority } from "../store/issues.js"
import {
  CreateIssueSchema,
  UpdateIssueSchema,
  ReorderSchema,
  CreateArtifactSchema,
  SaveArtifactContentSchema,
  validateBody,
} from "../lib/validation.js"
import {
  serializeIssue,
  serializePaginatedIssues,
  serializeIssueArtifact,
  serializeIssueArtifactContent,
} from "../lib/serialize.js"

const router = new Hono()

// GET /api/issues — paginated listing with filters
router.get("/", (c) => {
  const url = new URL(c.req.url)

  const page = Math.max(1, parseInt(url.searchParams.get("page") ?? "1", 10) || 1)
  const limit = Math.max(1, Math.min(100, parseInt(url.searchParams.get("limit") ?? "25", 10) || 25))

  // Filters
  const statusParam = url.searchParams.get("status")
  const status = statusParam
    ? statusParam.split(",").filter((s): s is IssueStatus =>
        ["open", "in-progress", "closed"].includes(s))
    : undefined

  const typeParam = url.searchParams.get("type")
  const type = typeParam
    ? typeParam.split(",").filter((t): t is IssueType =>
        ["bug", "task", "improvement", "chore"].includes(t))
    : undefined

  const priorityParam = url.searchParams.get("priority")
  const priority = priorityParam
    ? priorityParam.split(",").filter((p): p is Priority =>
        ["critical", "high", "medium", "low"].includes(p))
    : undefined

  const assigneeParam = url.searchParams.get("assignee")
  let assignee: string | null | undefined = undefined
  if (assigneeParam !== null) {
    assignee = assigneeParam === "__unassigned__" ? null : assigneeParam
  }

  const featureParam = url.searchParams.get("feature")
  let feature: string | null | undefined = undefined
  if (featureParam !== null) {
    feature = featureParam === "__unlinked__" ? null : featureParam
  }

  const pointsMinParam = url.searchParams.get("points_min")
  const pointsMin = pointsMinParam ? Number(pointsMinParam) : undefined

  const pointsMaxParam = url.searchParams.get("points_max")
  const pointsMax = pointsMaxParam ? Number(pointsMaxParam) : undefined

  const labelsParam = url.searchParams.get("labels")
  const labels = labelsParam ? labelsParam.split(",").filter(Boolean) : undefined

  // Sorting
  const sortParam = url.searchParams.get("sort") ?? "order"
  const sort = (["order", "title", "status", "type", "priority", "modified"].includes(sortParam)
    ? sortParam
    : "order") as "order" | "title" | "status" | "type" | "priority" | "modified"

  const dirParam = url.searchParams.get("dir") ?? "asc"
  const dir = dirParam === "desc" ? "desc" : "asc"

  const result = listIssues({
    page,
    limit,
    status,
    type,
    priority,
    assignee,
    feature,
    pointsMin: pointsMin !== undefined && !isNaN(pointsMin) ? pointsMin : undefined,
    pointsMax: pointsMax !== undefined && !isNaN(pointsMax) ? pointsMax : undefined,
    labels,
    sort,
    dir,
  })

  return c.json(serializePaginatedIssues(result))
})

// PATCH /api/issues/reorder — reorder issues by slug array
router.patch("/reorder", async (c) => {
  const body = await c.req.json()
  const parsed = validateBody(ReorderSchema, body)
  if (!parsed.success) {
    return c.json({ error: parsed.error }, 422)
  }
  const offset = parsed.data.offset ?? 0
  reorderIssues(parsed.data.slugs, offset)
  return c.json({ ok: true })
})

// GET /api/issues/:slug
router.get("/:slug", (c) => {
  const slug = c.req.param("slug")
  const issue = getIssue(slug)
  if (!issue) {
    return c.json({ error: "Issue not found", slug }, 404)
  }
  return c.json(serializeIssue(issue))
})

// POST /api/issues
router.post("/", async (c) => {
  const body = await c.req.json()
  const parsed = validateBody(CreateIssueSchema, body)
  if (!parsed.success) {
    return c.json({ error: parsed.error }, 422)
  }

  try {
    const issue = createIssue({
      title: parsed.data.title,
      body: parsed.data.body,
      type: parsed.data.type,
      priority: parsed.data.priority,
      feature: parsed.data.feature,
    })

    return c.json(serializeIssue(issue), 201)
  } catch (err) {
    const message = err instanceof Error ? err.message : "Failed to create issue"
    if (message.includes("already exists")) {
      return c.json({ error: message }, 409)
    }
    return c.json({ error: message }, 500)
  }
})

// PUT /api/issues/:slug
router.put("/:slug", async (c) => {
  const slug = c.req.param("slug")
  const body = await c.req.json()
  const parsed = validateBody(UpdateIssueSchema, body)
  if (!parsed.success) {
    return c.json({ error: parsed.error }, 422)
  }

  try {
    const issue = updateIssue(slug, {
      title: parsed.data.title,
      status: parsed.data.status,
      type: parsed.data.type,
      priority: parsed.data.priority,
      assignees: parsed.data.assignees,
      points: parsed.data.points,
      sprint: parsed.data.sprint,
      feature: parsed.data.feature,
      labels: parsed.data.labels,
      body: parsed.data.body,
    })

    if (!issue) {
      return c.json({ error: "Issue not found", slug }, 404)
    }

    return c.json(serializeIssue(issue))
  } catch (err) {
    const message = err instanceof Error ? err.message : "Failed to update issue"
    return c.json({ error: message }, 500)
  }
})

// DELETE /api/issues/:slug
router.delete("/:slug", (c) => {
  const slug = c.req.param("slug")
  const deleted = deleteIssue(slug)
  if (!deleted) {
    return c.json({ error: "Issue not found", slug }, 404)
  }
  return c.json({ ok: true, slug })
})

// GET /api/issues/:slug/artifacts
router.get("/:slug/artifacts", (c) => {
  const slug = c.req.param("slug")
  const artifacts = listIssueArtifacts(slug)
  if (artifacts === null) {
    return c.json({ error: "Issue not found", slug }, 404)
  }
  return c.json(artifacts.map(serializeIssueArtifact))
})

// POST /api/issues/:slug/artifacts/upload — upload a file
router.post("/:slug/artifacts/upload", async (c) => {
  const slug = c.req.param("slug")
  const formData = await c.req.formData()
  const file = formData.get("file")

  if (!file || !(file instanceof File)) {
    return c.json({ error: "file is required" }, 400)
  }

  const filename = file.name
  if (!filename || filename === "ISSUE.md") {
    return c.json({ error: "Invalid filename" }, 400)
  }

  try {
    const arrayBuffer = await file.arrayBuffer()
    const buffer = Buffer.from(arrayBuffer)
    const artifact = saveArtifactBuffer(slug, filename, buffer)
    if (!artifact) {
      return c.json({ error: "Issue not found", slug }, 404)
    }
    return c.json(serializeIssueArtifact(artifact), 201)
  } catch (err) {
    const message = err instanceof Error ? err.message : "Failed to upload file"
    return c.json({ error: message }, 500)
  }
})

// POST /api/issues/:slug/artifacts/create — create a new empty text file
router.post("/:slug/artifacts/create", async (c) => {
  const slug = c.req.param("slug")
  const body = await c.req.json()
  const parsed = validateBody(CreateArtifactSchema, body)
  if (!parsed.success) {
    return c.json({ error: parsed.error }, 422)
  }

  const filename = parsed.data.filename
  if (filename === "ISSUE.md") {
    return c.json({ error: "Invalid filename" }, 400)
  }

  // Check if file already exists
  const existing = getArtifactFilePath(slug, filename)
  if (existing) {
    return c.json({ error: "File already exists", filename }, 409)
  }

  try {
    const artifact = saveArtifactContent(slug, filename, "")
    if (!artifact) {
      return c.json({ error: "Issue not found", slug }, 404)
    }
    return c.json(serializeIssueArtifact(artifact), 201)
  } catch (err) {
    const message = err instanceof Error ? err.message : "Failed to create file"
    return c.json({ error: message }, 500)
  }
})

// GET /api/issues/:slug/artifacts/:filename — get artifact content
router.get("/:slug/artifacts/:filename", (c) => {
  const slug = c.req.param("slug")
  const filename = c.req.param("filename")
  const url = new URL(c.req.url)
  const raw = url.searchParams.has("raw")

  // If ?raw or binary file, serve the raw file
  if (raw || !isTextArtifact(filename)) {
    const result = getArtifactFilePath(slug, filename)
    if (!result) {
      return c.json({ error: "Artifact not found", slug, filename }, 404)
    }
    const content = readFileSync(result.filePath)
    return new Response(content, {
      headers: {
        "Content-Type": result.mimeType,
        "Content-Length": String(content.length),
        "Cache-Control": "no-cache",
      },
    })
  }

  // Text file — return as JSON
  const artifact = getArtifactContent(slug, filename)
  if (!artifact) {
    return c.json({ error: "Artifact not found", slug, filename }, 404)
  }
  return c.json(serializeIssueArtifactContent(artifact))
})

// PUT /api/issues/:slug/artifacts/:filename — save artifact content
router.put("/:slug/artifacts/:filename", async (c) => {
  const slug = c.req.param("slug")
  const filename = c.req.param("filename")
  const body = await c.req.json()
  const parsed = validateBody(SaveArtifactContentSchema, body)
  if (!parsed.success) {
    return c.json({ error: parsed.error }, 422)
  }

  const artifact = saveArtifactContent(slug, filename, parsed.data.content)
  if (!artifact) {
    return c.json({ error: "Issue not found or invalid filename", slug, filename }, 404)
  }
  return c.json(serializeIssueArtifact(artifact))
})

// DELETE /api/issues/:slug/artifacts/:filename — delete an artifact
router.delete("/:slug/artifacts/:filename", (c) => {
  const slug = c.req.param("slug")
  const filename = c.req.param("filename")
  const deleted = deleteArtifact(slug, filename)
  if (!deleted) {
    return c.json({ error: "Artifact not found", slug, filename }, 404)
  }
  return c.json({ ok: true, slug, filename })
})

export default router
